File: app/src/main/java/com/example/appctruyn/auth/AuthManager.java
package com.example.appctruyn.auth;

import com.example.appctruyn.model.User;
import com.google.android.gms.tasks.Task;
import com.google.android.gms.tasks.Tasks;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QuerySnapshot;

import java.util.List;

public class AuthManager {

    private static final FirebaseAuth auth = FirebaseAuth.getInstance();
    private static final FirebaseFirestore db = FirebaseFirestore.getInstance();

    public static FirebaseUser getCurrentUser() {
        return auth.getCurrentUser();
    }

    public static boolean isLoggedIn() {
        return auth.getCurrentUser() != null;
    }

    // Đăng ký
    public static Task<User> register(String email, String password, String displayName, String role) {
        return auth.createUserWithEmailAndPassword(email, password)
                .continueWithTask(task -> {
                    if (!task.isSuccessful()) throw task.getException();
                    
                    String uid = task.getResult().getUser().getUid();
                    User user = new User(uid, email, displayName, role);
                    
                    return db.collection("users").document(uid).set(user)
                            .continueWith(setTask -> {
                                if (!setTask.isSuccessful()) throw setTask.getException();
                                return user;
                            });
                });
    }

    // Đăng nhập
    public static Task<User> login(String email, String password) {
        return auth.signInWithEmailAndPassword(email, password)
                .continueWithTask(task -> {
                    if (!task.isSuccessful()) throw task.getException();
                    
                    String uid = task.getResult().getUser().getUid();
                    return db.collection("users").document(uid).get()
                            .continueWith(getTask -> {
                                if (!getTask.isSuccessful()) throw getTask.getException();
                                DocumentSnapshot doc = getTask.getResult();
                                User user = doc.toObject(User.class);
                                if (user == null) throw new Exception("Không tìm thấy thông tin người dùng");
                                return user;
                            });
                });
    }

    public static Task<User> getCurrentUserInfo() {
        FirebaseUser firebaseUser = getCurrentUser();
        if (firebaseUser == null) return Tasks.forResult(null);
        
        return db.collection("users").document(firebaseUser.getUid()).get()
                .continueWith(task -> {
                    if (!task.isSuccessful()) return null;
                    return task.getResult().toObject(User.class);
                });
    }

    // CẬP NHẬT ẢNH DƯỚI DẠNG BASE64
    public static Task<String> updateAvatarBase64(String base64String) {
        FirebaseUser firebaseUser = getCurrentUser();
        if (firebaseUser == null) return Tasks.forException(new Exception("Chưa đăng nhập"));
        
        return db.collection("users").document(firebaseUser.getUid())
                .update("avatarUrl", base64String)
                .continueWith(task -> {
                    if (!task.isSuccessful()) throw task.getException();
                    return base64String;
                });
    }

    // Admin: Lấy danh sách người dùng
    public static Task<List<User>> getAllUsers() {
        return db.collection("users").orderBy("createdAt").get()
                .continueWith(task -> {
                    if (!task.isSuccessful()) throw task.getException();
                    return task.getResult().toObjects(User.class);
                });
    }

    // Admin: Cập nhật role
    public static Task<Void> updateUserRole(String uid, String newRole) {
        return db.collection("users").document(uid).update("role", newRole);
    }

    public static void logout() {
        auth.signOut();
    }
}

File: app/src/main/java/com/example/appctruyn/auth/AuthManager.kt
package com.example.appctruyn.auth

import com.example.appctruyn.model.User
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.firestore.ktx.firestore
import com.google.firebase.ktx.Firebase
import kotlinx.coroutines.tasks.await

object AuthManager {

    private val auth = FirebaseAuth.getInstance()
    private val db = Firebase.firestore

    val currentUser: FirebaseUser? get() = auth.currentUser
    val isLoggedIn: Boolean get() = auth.currentUser != null

    // Đăng ký
    suspend fun register(email: String, password: String, displayName: String, role: String = "reader"): Result<User> {
        return try {
            val result = auth.createUserWithEmailAndPassword(email, password).await()
            val uid = result.user?.uid ?: throw Exception("Không thể tạo tài khoản")
            val user = User(uid = uid, email = email, displayName = displayName, role = role)
            db.collection("users").document(uid).set(user).await()
            Result.success(user)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // Đăng nhập
    suspend fun login(email: String, password: String): Result<User> {
        return try {
            val result = auth.signInWithEmailAndPassword(email, password).await()
            val uid = result.user?.uid ?: throw Exception("Đăng nhập thất bại")
            val userDoc = db.collection("users").document(uid).get().await()
            val user = userDoc.toObject(User::class.java) ?: throw Exception("Không tìm thấy thông tin")
            Result.success(user)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun getCurrentUserInfo(): User? {
        val uid = currentUser?.uid ?: return null
        return try {
            db.collection("users").document(uid).get().await().toObject(User::class.java)
        } catch (e: Exception) {
            null
        }
    }

    // CẬP NHẬT ẢNH DƯỚI DẠNG BASE64 (Không cần Firebase Storage)
    suspend fun updateAvatarBase64(base64String: String): Result<String> {
        val uid = currentUser?.uid ?: return Result.failure(Exception("Chưa đăng nhập"))
        return try {
            db.collection("users").document(uid).update("avatarUrl", base64String).await()
            Result.success(base64String)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // Admin: Lấy danh sách người dùng
    suspend fun getAllUsers(): Result<List<User>> {
        return try {
            val snapshot = db.collection("users").orderBy("createdAt").get().await()
            Result.success(snapshot.toObjects(User::class.java))
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // Admin: Cập nhật role
    suspend fun updateUserRole(uid: String, newRole: String): Result<Unit> {
        return try {
            db.collection("users").document(uid).update("role", newRole).await()
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    fun logout() { auth.signOut() }
}
